Add moving_avg pipeline aggregation. (#9657)
opensearch-project/documentation-website: _aggregations/pipeline/moving-avg.md

---
layout: default
title: Moving average
parent: Pipeline aggregations
nav_order: 120
---

## Moving average aggregations

The `moving_avg` aggregation has been deprecated in favor of the `moving_fn` aggregation.
{: .important}

A `moving_avg` aggregation is a parent pipeline aggregation that calculates a sequence of averages of a metric contained in windows (adjacent subsets) of an ordered dataset. 

To create a `moving_avg` aggregation you first create a `histogram` or `date_histogram` aggregation. Optionally, you then embed a metric aggregation in the histogram aggregation. Finally, embed the `moving_avg` aggreation in the histogram and set the `buckets_path` parameter to the embedded metric that you want to track. 

> A window's size is the number of sequential data values in the window. On each iteration, the algorithm calculates the average for all data points in the window and then slides forward one data value, excluding the first value of the previous window and including the first value of the next window.

Reviewer: "For" is not the meaning that we are going for here. If not "on", then maybe "at" or "during"?

| `buckets_path`        | Required          | String          | The path of the buckets to be aggregated. See [Pipeline aggregations]({{site.url}}{{site.baseurl}}/aggregations/pipeline/index#buckets-path). |
| `gap_policy`          | Optional          | String          | The policy to apply to missing data. Valid values are `skip`, `insert_zeros`, and `keep_values`. Default is `skip`. |
| `format`              | Optional          | String          | A [DecimalFormat](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/text/DecimalFormat.html) formatting string. Returns the formatted output in the aggregation's `value_as_string` property. |
| `window`              | Optional          | Numerical       | The number of data points contained in the window. Default is `5`. |
| `model`               | Optional          | String          | The weighted moving average model to use. Options are `ewma`, `holt`, `holt_winters`, `linear`, and `simple`. Default is `simple`. See [Models](#models). |
| `settings`            | Optional          | Object          | Parameters for adjusting the window. See [Models](#models). |

> | `predict`             | Optional          | Numeical        | The number of predicted values to append to the end of the result. Default is `0`. |

Reviewer: ```suggestion
| `predict`             | Optional          | Numerical        | The number of predicted values to append to the end of the result. Default is `0`. |
```

            "value": 8297293.6
          }
        },
        {
          "key_as_string": "2025-06-23T00:00:00.000Z",
          "key": 1750636800000,
          "doc_count": 0,
          "moving_avg_of_sum_of_bytes": {
            "value": 8297293.6
          }
        },
        {
          "key_as_string": "2025-06-30T00:00:00.000Z",
          "key": 1751241600000,
          "doc_count": 0,
          "moving_avg_of_sum_of_bytes": {
            "value": 8297293.6
          }
        }
      ]
    }
  }
}
```
</details>

## Models

The `moving_avg` aggregation supports five models that differ in how they weight values in the moving window. 

Use the `model` parameter to specify which model to use.

| Model | Model keyword | Weighting   |
|-------|---------------|-------------|
| Simple | `simple` | An unweighted mean of all values in the window. |
| Linear | `linear` | Uses a linear decay of weights, giving more importance to recent values. |
| Exponentially Weighted Moving Average | `ewma` | Uses exponentially decreasing weights, giving more importance to recent values. |
| Holt | `holt` | Uses a second exponential term to smooth long-term trends. |

> | Holt-Winters | `holt_winters` | Uses a third exponential term to smooth perioic (seasonal) effects. |

Reviewer 2: ```suggestion
| Holt-Winters | `holt_winters` | Uses a third exponential term to smooth periodic (seasonal) effects. |
```

> Use the `settings` object to set the model's properties. The following table shows the settings available for each model.

Reviewer 2: ```suggestion
Use the `settings` object to set the model's properties. The following table shows the available settings for each model.
```

> | Model            | Parameter   | Allowed values  | Default | Description |

Reviewer 3: Similar minor comment on table visulaization: https://github.com/opensearch-project/documentation-website/pull/9796#discussion_r2098653755

<table>
  <thead>
    <tr>
      <th>Model</th>
      <th>Parameter</th>
      <th>Allowed values</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>simple</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>linear</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>ewma</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="2" style="vertical-align: middle;"><code>holt</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.1</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="6" style="vertical-align: middle;"><code>holt_winters</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td><code>gamma</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The periodic decay parameter.</td>
    </tr>
    <tr>
      <td><code>type</code></td>
      <td><code>add</code>, <code>mult</code></td>
      <td><code>add</code></td>
      <td>How seasonality is incorporated.</td>
    </tr>
    <tr>
      <td><code>period</code></td>
      <td>Integer</td>
      <td>1</td>
      <td>The number of buckets that comprise the period.</td>
    </tr>
    <tr>
      <td><code>pad</code></td>
      <td>Boolean</td>
      <td>true</td>
      <td>Whether to add a small offset to <code>0</code> values for <code>mult</code> type models to avoid a divide-by-zero error.</td>
    </tr>
  </tbody>
</table>

```
<table>
  <thead>
    <tr>
      <th>Model</th>
      <th>Parameter</th>
      <th>Allowed values</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>simple</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>linear</code></td>
      <td>None</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>ewma</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="2" style="vertical-align: middle;"><code>holt</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.1</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td rowspan="6" style="vertical-align: middle;"><code>holt_winters</code></td>
      <td><code>alpha</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The decay parameter.</td>
    </tr>
    <tr>
      <td><code>beta</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The trend decay parameter.</td>
    </tr>
    <tr>
      <td><code>gamma</code></td>
      <td>[0, 1]</td>
      <td>0.3</td>
      <td>The periodic decay parameter.</td>
    </tr>
    <tr>
      <td><code>type</code></td>
      <td><code>add</code>, <code>mult</code></td>
      <td><code>add</code></td>
      <td>How seasonality is incorporated.</td>
    </tr>
    <tr>
      <td><code>period</code></td>
      <td>Integer</td>
      <td>1</td>
      <td>The number of buckets that comprise the period.</td>
    </tr>
    <tr>
      <td><code>pad</code></td>
      <td>Boolean</td>
      <td>true</td>
      <td>Whether to add a small offset to <code>0</code> values for <code>mult</code> type models to avoid a divide-by-zero error.</td>
    </tr>
  </tbody>
</table>
```

Reviewer: Reformatted the table in the same way as moving_fn.

| `holt`              | `alpha`            | [0, 1] | 0.3     | The decay parameter for the level component. |
|              | `beta`             | [0, 1] | 0.1     | The decay parameter for the trend component.|
| `holt_winters`      | `alpha`            | [0, 1] | 0.3     | The decay parameter for the level component.  |
|       | `beta`             | [0, 1] | 0.3     | The decay parameter for the trend component. |
|       | `gamma`            | [0, 1] | 0.3     | The decay parameter for the seasonal component.  |
|       | `type`             | `add`, `mult`   | `add`   | Defines how seasonality is modeled: additive or multiplicative. |
|       | `period`           | Integer         | 1       | The number of buckets comprising the period. |
|      | `pad`              | Boolean         | `true`    | Whether to add a small offset to `0` values for `mult` type models to avoid a divide-by-zero error. |

For a discussion of these models and their parameters, see [Wikipedia](https://en.wikipedia.org/wiki/Moving_average). 

### Example: Holt model

The `holt` model computes weights with exponential decay controlled by the `alpha` and `beta` parameters.

> The following calculates a moving average against weekly byte total data using a Holt model with a `window` size of `6`, an `alpha` value of `0.4`, and a `beta` value of `0.2`:

Reviewer 2: ```suggestion
The following request calculates a moving average against total weekly byte data using a Holt model with a `window` size of `6`, an `alpha` value of `0.4`, and a `beta` value of `0.2`:
```

> 

Reviewer 2: Line 393: "against" =>"of"?

---
layout: default
title: Moving average
parent: Pipeline aggregations
nav_order: 120
redirect_from:
  - /query-dsl/aggregations/pipeline-agg#moving_avg/
  - /query-dsl/aggregations/pipeline/moving-avg/
---

## Moving average aggregations

> A `moving_avg` aggregation is a parent pipeline aggregation that calculates a sequence of averages of a metric contained in windows (adjacent subsets) of an ordered dataset. 

Reviewer 3: From code:

"The moving_avg aggregation has been deprecated in favor of the moving_fn aggregation."

---------

Since this seems to be on deprecation path, I think it might be better to document `moving_fn` instead.

Author: If the functionality is still in the product, it should be documented, even if it's deprecated.

Deprecation notices should be included in the release notes. It can also be mentioned in the doc for the deprecated functionality if that's documentation policy. 

@kolchfa-aws , should this page contain a deprecation notice?

I'm in the process of also documenting the moving_fn aggregation; a PR should be available soon.

Reviewer: @dwelsch-esi Yes, there should be a deprecation notice for this one, advising to use moving_fn aggregation instead.

Reviewer: Added a deprecation notice.

For example, given the data `[1, 5, 8, 23, 34, 28, 7, 23, 20, 19]`, a moving average with a window size of 5 is as follows:

```
(1 + 5 + 8 + 23 + 34) / 5 = 14.2
(5 + 8 + 23 + 34 + 28) / 5 = 19.6
(8 + 23 + 34 + 28 + 7) / 5 = 20
and so on ...
```

The `moving_avg` aggregation is typically applied to time-series data to smooth out noise or short-term fluctuations and to identify trends.
Specify a small window size to smooth out small-scale fluctuations. Specify a larger window size  to smooth out high-frequency fluctuations or random noise, making lower frequency trends more visible.

For more information about moving averages, see [Wikipedia](https://en.wikipedia.org/wiki/Moving_average).

## Parameters

The `moving_avg` aggregation takes the following parameters.

| Parameter             | Required/Optional | Data type       | Description |
| :--                   | :--               |  :--            | :--         |

> | `buckets_path`        | Required          | String          | The path of the buckets to be aggregated. See [Pipeline aggregations]({{site.url}}{{site.baseurl}}/aggregations/pipeline/index#pipeline-aggregation-syntax). |

Author: ```suggestion
| `buckets_path`        | Required          | String          | The path of the buckets to be aggregated. See [Pipeline aggregations]({{site.url}}{{site.baseurl}}/aggregations/pipeline/index#buckets-path-property). |
```
